extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/oauth/OAuth2IdentityManager.java

```diff
@@ -0,0 +1,61 @@
+package io.quarkus.elytron.security.oauth2.runtime.oauth;
+
+import org.jboss.logging.Logger;
+import org.wildfly.security.auth.server.RealmUnavailableException;
+import org.wildfly.security.auth.server.SecurityDomain;
+import org.wildfly.security.auth.server.SecurityIdentity;
+import org.wildfly.security.evidence.BearerTokenEvidence;
+
+import io.undertow.security.idm.Account;
+import io.undertow.security.idm.Credential;
+import io.undertow.security.idm.IdentityManager;
+
+public class OAuth2IdentityManager implements IdentityManager {
+    private static Logger log = Logger.getLogger(OAuth2IdentityManager.class);
+    private final SecurityDomain domain;
+
+    public OAuth2IdentityManager(SecurityDomain domain) {
+        this.domain = domain;
+    }
+
+    @Override
+    public Account verify(Account account) {
+        return account;
+    }
+
+    @Override
+    public Account verify(String id, Credential credential) {
+        return verify(null, credential);
+    }
+
+    @Override
+    public Account verify(Credential credential) {
+        log.debugf("verify, credential=%s", credential);
+        try {
+            if (credential instanceof Oauth2Credential) {
+                Oauth2Credential oauth2Credential = (Oauth2Credential) credential;
+
+                try {
+                    BearerTokenEvidence evidence = new BearerTokenEvidence(oauth2Credential.getBearerToken());
+                    SecurityIdentity result = domain.authenticate(evidence);
+                    //FIXME unable to find where I should read the roles from so I did it here
+                    ElytronOAuth2CallerPrincipal principal = (ElytronOAuth2CallerPrincipal) result.getPrincipal();
```

Comment:
I agree we should parse roles differently.

The `SecurityIdentity` already provides a `getRoles` method. This method returns the roles based on the `RoleDecoder` configured to a realm. The role decoder is responsible to decode roles from the attributes (in this case the claims from the token) associated with the identity from the realm. That said, we should probably allow developers to inject a custom role decoder and be more flexible in regards to claims to roles mapping.

Reply:
@pedroigor Hi, thanks this is interesting. FYI, in smallrye-jwt we have a property such as `smallrye.jwt.path.groups` (which in this can be used to map the roles info to the `groups` claim expected by MP JWT), example, "realm_acess/roles", we used a KC token to test it :-), which can be handy for the tokens issued by KC before it could also create a top level groups claim itself :-)
@loicmathieu can you please experiment with the `SecurityIdentity.getRoles` and/or `RolesDecoder` as recommended by @pedroigor and see if you can the `scopes` values ?

Reply:
Hi,

I will add two configuration option, one for the claim name and the other one for the role token separator (empty will means that the claim is an array).

I will check the `SecurityIdentity.geRoles` method and the `RoleDecoder` to implement the role mapping properly, 

I'm very interesting to know how we can provides a way to a user to provide it's own implementation of a `RoleDecoder`. A I understand it,  with Quarkus stuff are done at build time so we can't just check that a `RoleDecoder` implementation is provided at runtime right ? Or maybe I'm wrong and I can just check in ARC bean container if a `RoleDecoder` exist at build time (STATIT_INIT phase ?) and if it exist use it otherwise use the default one from the framework.

Reply:
@loicmathieu 
What do you mean under a role token separator ? in smallrye-jwt we use a simple convention, it can be "a" or "a/b" (up to a max depth of 4). Every segment but the last one must be a JSON object while the last one - depends on type of the claim, if it is roles related - must either be an array or a single simple value (we don't support the latter).
But may be this kind flexibility is not immediately required if the default RoleDecoder uses `scope` ?

Reply:
@sberyozkin in an introspection endpoint, the scopes are separated by a space inside the same attribute. So I use a roleSeparator (by default a space) to split the attribute to get the multple roles from the attribute. If there is no separator I assume the attribute is an array so it can covers multiple usecase : 
- an attribute with a separated list of roles (match the scope attribute with a space separator by default)
- an attribute that is a JSON array

Maybe not everything should be documented now if there is a refactoring/consolidation forseen on the different security implementation we can find a common way of configuring the role decoding later.

Current attribute implementation in the wildfly bearer token realm don't allow hierarchical attribute (like userinfo.roles), I think this is more an OIDC use case ...

For the `RoleDecoded` I didn't find how to configure it, it seems to comes from the `SecurityDomain` that itself takes it from the `RealmInfo` but I don't know how to set such `RealmInfo` ... help will be appreciated ;)

Reply:
@loicmathieu sorry, yeah, ignore my comment above, I forgot it was a single value (the introspection response being JSON, they shoud've spec-ed it better really, it does not need to be URL safe there).

That is fine. Re `RoleDecoder`, may be it can be customized in the next phase, @pedroigor @sebastienblanc would it be OK ?

Reply:
Yeah, for me is fine. I think that to get this forward we would need the `elytron-security` extension providing this capability.

The configuration option for the claim name should be enough for now.

Reply:
@pedroigor your PR to get it addressed at the `elytron-security` level looks nice. I guess once it is in, the code which @loicmathieu did for checking the claim properties can be wrapped into this extension specific RoleDecoder :-). May be some ordering constraints can be applied in time as well

Reply:
Yeah, that is the idea. So that you just rely on the roles available through a `SecurityIdentity.getRoles` call.

Elytron is very flexible in regards to role mapping and in the future we could also allow apps to customized how roles are actually mapped by using a `RoleMapper`.